src/util/HashSet.h

```diff
@@ -32,4 +54,142 @@ template <class T,
 using HashSetWithMemoryLimit =
     absl::flat_hash_set<T, HashFct, EqualElem, Alloc>;
 
+// Wrapper around absl::node_hash_set with a memory limit. All operations that
+// may change the allocated memory of the hash set is tracked using a
+// `AllocationMemoryLeftThreadsafe` object.
+template <class T, class SizeGetter = SizeOfSizeGetter,
+          class HashFct = absl::container_internal::hash_default_hash<T>,
+          class EqualElem = absl::container_internal::hash_default_eq<T>>
+class CustomHashSetWithMemoryLimit {
+ private:
+  using HashSet = absl::node_hash_set<T, HashFct, EqualElem>;
+  HashSet hashSet_;
+  detail::AllocationMemoryLeftThreadsafe memoryLeft_;
+  MemorySize memoryUsed_;
+  SizeGetter sizeGetter_;
+  size_t currentSlotSize_;
+
+ public:
+  CustomHashSetWithMemoryLimit(
+      detail::AllocationMemoryLeftThreadsafe memoryLeft,
+      SizeGetter sizeGetter = {})
+      : memoryLeft_{memoryLeft},
+        memoryUsed_{MemorySize::bytes(0)},
+        sizeGetter_{sizeGetter},
+        currentSlotSize_{0} {
+    // Once the hash set is initialized, calculate the initial memory
+    // used by the slots of the hash set
+    updateSlotArrayMemoryUsage();
+  }
+
+  ~CustomHashSetWithMemoryLimit() { decreaseMemoryUsed(memoryUsed_); }
+
+  // Try to allocate the amount of memory requested
+  void increaseMemoryUsed(ad_utility::MemorySize amount) {
+    memoryLeft_.ptr()->wlock()->decrease_if_enough_left_or_throw(amount);
```

Comment:
This is a little wasteful, as we always have to obtain a global mutex.
I think for a follow-up (or preparation) PR, 
You could implement a wrapper around the memoryLeftThreadsafe object, that stores a (single-threaded) small pool of memory and only goes to the global `wlock()` when that pool is exhausted.
Currently `increase (1), increase(1), increase(1)` needs a global synchronization for each of the three inserts, which seems wastefult. 
(But as you have abstracted away all the memory handling, that is easy to integrate.)